### PR TITLE
Use cython fortran-record reader

### PR DESCRIPTION
**Pull request by cphyc**
This changes the way Ramses is being read from a python-based implementation to a Cython-based one. The source code can be found here https://github.com/cphyc/cython_fortran_file and could be included in pynbody if required.

Here is a short summary of the performance changes from the python implementation (current) to the cython implementation (this PR).

| Quantity | Python-based access time | Cython-based access time | ratio | 
| ----------- | ----------------------------------- | ----------------------------------- | -- | 
| dark matter position |  0.037s                           | 0.043s                  | 1.16 |
| gas density |  0.410s                           | 0.148s                               | 0.36 |

The faster access for gas density is due to the IO being dominated by skipping most of the records in the file, which is much faster when done in Cython than in Python. For particle-based access, the IO is mostly bound by reading speeds from disk, with a slight benefit for the python-based version (which is based on a memory map).

### CODE REVIEW

**Comment by apontzen** [comment]
The gains seem to be marginal, given the load times were already looking acceptable, although I wonder what size of file you're testing it on? Also `cython_fortran_file` installation is failing on travis.

Overall I'm a little ambivalent about whether this update is worth it, unless perhaps it produces bigger speed-ups for bigger/more complex files.

**Comment by cphyc** [comment]
This is for a hydrodynamical simulation. The times are obtained using the code below (in IPython). All runs are performed on the same computer twice.

|                                       | Load Time(s) (`.load`) | Particle reading time (s) | Hydro reading time (s) | 
|--------------------------------|-------------------------------|----------------------------------|-------------------------------|
| yt                                   | ~13.3s   (`ds.index`)    | ~7.4s                              | ~12s                            |
| pynbody, Python reader | ~1min50s-1min57s    |  ~1.6s-1.9s                    |   14min44s                   |
| pynbody, Cython reader |  ~6.20s                        |  ~2.2s                            | ~23s                            |     

It seems that compared to the small simulations, the loading times are much improved with the Cython version for the hydro and the AMR (as only small records are being red). I don't think we should worry about the small regression for particle reading.

---
For pynbody
```python
import pynbody

f = pynbody.load('output_00163')
%time f = pynbody.load('output_00163')
f.dm['iord']

f = pynbody.load('output_00163')
%time f.dm['iord']
f.gas['rho']

f = pynbody.load('output_00163')
%time f.gas['rho']
```
For yt
```python
import yt
ds = yt.load('output_00163/info_00163.txt'); ds.index
ds = yt.load('output_00163/info_00163.txt')
%time ds.index
ds.r['io','particle_identity']

ds = yt.load('output_00163/info_00163.txt'); ds.index
%time ds.r['io','particle_identity']
ds.r['gas', 'density']

ds = yt.load('output_00163/info_00163.txt'); ds.index
%time ds.r['gas','density']
```

**Comment by apontzen** [comment]
OK! I am convinced :-)

**Comment by cphyc** [comment]
For the sake of comparison, the same code is used internally in yt since https://github.com/yt-project/yt/pull/1771.
Similar speedups were reported (see e.g. https://github.com/yt-project/yt/pull/1771#issuecomment-385422501 )

**Comment by cphyc** [comment]
This turned out to be more complicated than I initially thought. @apontzen do you understand why the code fails on Python 2 but not on Python 3 ?
There is probably some magic involved with how packages are handled in the different Python versions, but I am no Python 2 expert.

**Comment by apontzen** [comment]
Not sure -- I just tried this branch on my own python2.7 installation and do not encounter the same problem... a bit stumped as to why travis sees this failure. Could it be something to do with the version of cython being installed? Here is what I got from a fresh anaconda install, not pinning any module versions:

```
# packages in environment at /Users/app/opt/anaconda3/envs/py27:
#
# Name                    Version                   Build  Channel
_anaconda_depends         2019.03                  py27_0  
alabaster                 0.7.12                   py27_0  
anaconda                  custom                   py27_1  
anaconda-client           1.7.2                    py27_0  
anaconda-project          0.8.4                      py_0  
applaunchservices         0.2.1                      py_0  
appnope                   0.1.0            py27hb466136_0  
appscript                 1.1.0            py27h1de35cc_0  
argh                      0.26.2                   py27_0  
asn1crypto                1.2.0                    py27_0  
astroid                   1.6.5                    py27_0  
astropy                   2.0.9            py27h1d22016_0  
atomicwrites              1.3.0                    py27_1  
attrs                     19.3.0                     py_0  
autopep8                  1.4.4                      py_0  
babel                     2.7.0                      py_0  
backports                 1.0                        py_2  
backports.functools_lru_cache 1.6.1                      py_0  
backports.os              0.1.1                    py27_0  
backports.shutil_get_terminal_size 1.0.0                    py27_2  
backports_abc             0.5              py27h6972548_0  
beautifulsoup4            4.8.1                    py27_0  
bitarray                  1.2.0            py27h1de35cc_0  
bkcharts                  0.2              py27haafc882_0  
blas                      1.0                         mkl  
bleach                    3.1.0                    py27_0  
blosc                     1.16.3               hd9629dc_0  
bokeh                     1.4.0                    py27_0  
boto                      2.49.0                   py27_0  
bottleneck                1.3.1            py27h1d22016_0  
bzip2                     1.0.8                h1de35cc_0  
ca-certificates           2019.11.27                    0  
cdecimal                  2.3              py27h1de35cc_3  
certifi                   2019.11.28               py27_0  
cffi                      1.13.2           py27hb5b8e2f_0  
chardet                   3.0.4                 py27_1003  
click                     7.0                      py27_0  
cloudpickle               1.2.2                      py_0  
clyent                    1.2.2                    py27_1  
colorama                  0.4.3                      py_0  
configparser              4.0.2                    py27_0  
contextlib2               0.6.0.post1                py_0  
cryptography              2.8              py27ha12b0ac_0  
curl                      7.67.0               ha441bb4_0  
cycler                    0.10.0           py27hfc73c78_0  
cython                    0.29.14          py27h0a44026_0  
cytoolz                   0.10.1           py27h1de35cc_0  
dask                      1.1.4                    py27_1  
dask-core                 1.1.4                    py27_1  
dbus                      1.13.12              h90a0687_0  
decorator                 4.4.1                      py_0  
defusedxml                0.6.0                      py_0  
diff-match-patch          20181111                   py_0  
distributed               1.28.1                   py27_0  
docutils                  0.15.2                   py27_0  
entrypoints               0.3                      py27_0  
enum34                    1.1.6                    py27_1  
et_xmlfile                1.0.1            py27hc42f929_0  
expat                     2.2.6                h0a44026_0  
fastcache                 1.1.0            py27h1de35cc_0  
filelock                  3.0.12                     py_0  
flake8                    3.7.9                    py27_0  
flask                     1.1.1                      py_0  
freetype                  2.9.1                hb4e5f40_0  
funcsigs                  1.0.2            py27hb9f6266_0  
functools32               3.2.3.2                  py27_1  
future                    0.18.2                   py27_0  
futures                   3.3.0                    py27_0  
get_terminal_size         1.0.0                h7520d66_0  
gettext                   0.19.8.1             h15daf44_3  
gevent                    1.4.0            py27h1de35cc_0  
glib                      2.63.1               hd977a24_0  
glob2                     0.7                        py_0  
gmp                       6.1.2                hb37e062_1  
gmpy2                     2.0.8            py27h6ef4df4_2  
greenlet                  0.4.15           py27h1de35cc_0  
grin                      1.2.1                    py27_4  
h5py                      2.9.0            py27h3134771_0  
hdf5                      1.10.4               hfa1e0ec_0  
heapdict                  1.0.1                      py_0  
html5lib                  1.0.1                    py27_0  
icu                       58.2                 h4b95b61_1  
idna                      2.8                      py27_0  
imageio                   2.6.1                    py27_0  
imagesize                 1.1.0                    py27_0  
importlib_metadata        1.3.0                    py27_0  
intel-openmp              2019.4                      233  
intervaltree              3.0.2                      py_0  
ipaddress                 1.0.23                     py_0  
ipykernel                 4.10.0                   py27_0  
ipython                   5.8.0                    py27_0  
ipython_genutils          0.2.0                    py27_0  
ipywidgets                7.5.1                      py_0  
isort                     4.3.21                   py27_0  
itsdangerous              1.1.0                    py27_0  
jbig                      2.1                  h4d881f8_0  
jdcal                     1.4.1                      py_0  
jedi                      0.14.1                   py27_0  
jinja2                    2.10.3                     py_0  
jpeg                      9b                   he5867d9_2  
jsonschema                3.2.0                    py27_0  
jupyter                   1.0.0                    py27_7  
jupyter_client            5.3.4                    py27_0  
jupyter_console           5.2.0                    py27_1  
jupyter_core              4.6.1                    py27_0  
jupyterlab                0.33.11                  py27_0  
jupyterlab_launcher       0.11.2           py27h28b3542_0  
keyring                   18.0.0                   py27_0  
kiwisolver                1.1.0            py27h0a44026_0  
krb5                      1.16.4               hddcf347_0  
lazy-object-proxy         1.4.3            py27h1de35cc_0  
libarchive                3.3.3                h786848e_5  
libcurl                   7.67.0               h051b688_0  
libcxx                    4.0.1                hcfea43d_1  
libcxxabi                 4.0.1                hcfea43d_1  
libedit                   3.1.20181209         hb402a30_0  
libffi                    3.2.1                h475c297_4  
libgfortran               3.0.1                h93005f0_2  
libiconv                  1.15                 hdd342a3_7  
liblief                   0.9.0                h2a1bed3_2  
libpng                    1.6.37               ha441bb4_0  
libsodium                 1.0.16               h3efe00b_0  
libspatialindex           1.9.3                h0a44026_0  
libssh2                   1.8.2                ha12b0ac_0  
libtiff                   4.1.0                hcb84e12_0  
libxml2                   2.9.9                hf6e021a_1  
libxslt                   1.1.33               h33a18ac_0  
linecache2                1.0.0                    py27_0  
llvmlite                  0.30.0           py27h98b8051_0  
locket                    0.2.0            py27ha10513d_1  
lxml                      4.4.2            py27hef8c89e_0  
lz4-c                     1.8.1.2              h1de35cc_0  
lzo                       2.10                 h362108e_2  
markupsafe                1.1.1            py27h1de35cc_0  
matplotlib                2.2.3            py27h54f8f79_0  
mccabe                    0.6.1                    py27_1  
mistune                   0.8.4            py27h1de35cc_0  
mkl                       2019.4                      233  
mkl-service               2.3.0            py27hfbe908c_0  
mkl_fft                   1.0.15           py27h5e564d8_0  
mkl_random                1.1.0            py27ha771720_0  
mock                      3.0.5                    py27_0  
more-itertools            5.0.0                    py27_0  
mpc                       1.1.0                h6ef4df4_1  
mpfr                      4.0.1                h3018a27_3  
mpmath                    1.1.0                    py27_0  
msgpack-python            0.6.1            py27h04f5b5a_1  
multipledispatch          0.6.0                    py27_0  
nbconvert                 5.6.1                    py27_0  
nbformat                  4.4.0                    py27_0  
ncurses                   6.1                  h0a44026_1  
networkx                  2.2                      py27_1  
nltk                      3.4.5                    py27_0  
nose                      1.3.7                    py27_2  
notebook                  5.7.8                    py27_0  
numba                     0.46.0           py27h6440ff4_0  
numexpr                   2.7.0            py27h7413580_0  
numpy                     1.16.5           py27hacdab7b_0  
numpy-base                1.16.5           py27h6575580_0  
numpydoc                  0.9.1                      py_0  
olefile                   0.46                     py27_0  
openpyxl                  2.6.1                    py27_1  
openssl                   1.1.1d               h1de35cc_3  
packaging                 19.2                       py_0  
pandas                    0.24.2           py27h0a44026_0  
pandoc                    2.2.3.2                       0  
pandocfilters             1.4.2                    py27_1  
parso                     0.5.2                      py_0  
partd                     0.3.10                   py27_1  
path.py                   11.5.0                   py27_0  
pathlib2                  2.3.5                    py27_0  
pathtools                 0.1.2                      py_1  
patsy                     0.5.1                    py27_0  
pcre                      8.43                 h0a44026_0  
pep8                      1.7.1                    py27_0  
pexpect                   4.7.0                    py27_0  
pickleshare               0.7.5                    py27_0  
pillow                    6.2.1            py27hb68e598_0  
pip                       19.3.1                   py27_0  
pkginfo                   1.5.0.1                  py27_0  
pluggy                    0.13.1                   py27_0  
ply                       3.11                     py27_0  
portaudio                 19.6.0               h41429eb_1  
posix-ipc                 1.0.4                    pypi_0    pypi
prometheus_client         0.7.1                      py_0  
prompt_toolkit            1.0.15           py27h4a7b9c2_0  
psutil                    5.6.7            py27h1de35cc_0  
ptyprocess                0.6.0                    py27_0  
py                        1.8.0                    py27_0  
py-lief                   0.9.0            py27h1413db1_2  
pyaudio                   0.2.11           py27h1de35cc_1  
pycodestyle               2.5.0                    py27_0  
pycosat                   0.6.3            py27h1de35cc_0  
pycparser                 2.19                     py27_0  
pycrypto                  2.6.1            py27h1de35cc_9  
pycurl                    7.43.0.3         py27ha12b0ac_0  
pydocstyle                3.0.0                    py27_0  
pyflakes                  2.1.1                    py27_0  
pygments                  2.5.2                      py_0  
pylint                    1.9.2                    py27_0  
pympler                   0.7                        py_0  
pynbody                   0.48                      dev_0    <develop>
pyodbc                    4.0.27           py27h0a44026_0  
pyopenssl                 19.1.0                   py27_0  
pyparsing                 2.4.5                      py_0  
pyqt                      5.9.2            py27h655552a_2  
pyrsistent                0.15.6           py27h1de35cc_0  
pysocks                   1.7.1                    py27_0  
pytables                  3.5.2            py27h5bccee9_1  
pytest                    4.6.2                    py27_0  
python                    2.7.17               h97142e2_0  
python-dateutil           2.8.1                      py_0  
python-jsonrpc-server     0.3.2                      py_0  
python-language-server    0.31.2                   py27_0  
python-libarchive-c       2.8                     py27_13  
python.app                2                        py27_9  
pytz                      2019.3                     py_0  
pywavelets                1.0.3            py27h1d22016_1  
pyyaml                    5.2              py27h1de35cc_0  
pyzmq                     18.1.0           py27h0a44026_0  
qdarkstyle                2.7                        py_0  
qt                        5.9.7                h468cd18_1  
qtawesome                 0.6.0                      py_0  
qtconsole                 4.6.0                      py_0  
qtpy                      1.9.0                      py_0  
readline                  7.0                  h1de35cc_5  
requests                  2.22.0                   py27_1  
rope                      0.14.0                     py_0  
rtree                     0.8.3                    py27_0  
ruamel_yaml               0.15.46          py27h1de35cc_0  
scandir                   1.10.0           py27h1de35cc_0  
scikit-image              0.14.2           py27h0a44026_0  
scikit-learn              0.20.3           py27h27c97d8_0  
scipy                     1.2.1            py27h1410ff5_0  
seaborn                   0.9.0              pyh91ea838_1  
selectors2                2.0.1                      py_0  
send2trash                1.5.0                    py27_0  
setuptools                42.0.2                   py27_0  
simplegeneric             0.8.1                    py27_2  
singledispatch            3.4.0.3          py27he22c18d_0  
sip                       4.19.8           py27h0a44026_0  
six                       1.13.0                   py27_0  
snappy                    1.1.7                he62c110_3  
snowballstemmer           2.0.0                      py_0  
sortedcollections         1.1.2                    py27_0  
sortedcontainers          2.1.0                    py27_0  
soupsieve                 1.9.5                    py27_0  
sphinx                    1.8.5                    py27_0  
sphinxcontrib             1.0                      py27_1  
sphinxcontrib-websupport  1.1.2                      py_0  
spyder                    4.0.0                    py27_0  
spyder-kernels            1.8.1                    py27_0  
sqlalchemy                1.3.11           py27h1de35cc_0  
sqlite                    3.30.1               ha441bb4_0  
ssl_match_hostname        3.7.0.1                  py27_0  
statsmodels               0.10.1           py27h1d22016_0  
subprocess32              3.5.4            py27h1de35cc_0  
sympy                     1.5                      py27_0  
tbb                       2019.8               h04f5b5a_0  
tblib                     1.6.0                      py_0  
terminado                 0.8.3                    py27_0  
testpath                  0.4.4                      py_0  
tk                        8.6.8                ha441bb4_0  
toolz                     0.10.0                     py_0  
tornado                   5.1.1            py27h1de35cc_0  
tqdm                      4.40.2                     py_0  
traceback2                1.4.0                    py27_0  
traitlets                 4.3.3                    py27_0  
typing                    3.7.4.1                  py27_0  
ujson                     1.35             py27h1de35cc_0  
unicodecsv                0.14.1           py27h170f95c_0  
unittest2                 1.1.0                    py27_0  
unixodbc                  2.3.7                h1de35cc_0  
urllib3                   1.25.7                   py27_0  
watchdog                  0.9.0                    py27_1  
wcwidth                   0.1.7                    py27_0  
webencodings              0.5.1                    py27_1  
werkzeug                  0.16.0                     py_0  
wheel                     0.33.6                   py27_0  
widgetsnbextension        3.5.1                    py27_0  
wrapt                     1.11.2           py27h1de35cc_0  
wurlitzer                 2.0.0                    py27_0  
xlrd                      1.2.0                    py27_0  
xlsxwriter                1.2.6                      py_0  
xlwings                   0.16.3                   py27_0  
xlwt                      1.2.0            py27hbeec4ae_0  
xz                        5.2.4                h1de35cc_4  
yaml                      0.1.7                hc338f04_2  
yapf                      0.28.0                     py_0  
zeromq                    4.3.1                h0a44026_3  
zict                      1.0.0                      py_0  
zipp                      0.6.0                      py_0  
zlib                      1.2.11               h1de35cc_3  
zstd                      1.3.7                h5bba6e5_0  
```

File: pynbody/snapshot/grafic.py [hunk 1]
```diff
@@ -196,27 +199,32 @@ def _read_pvar(self):
...
-                    re = f.get_raw_memmapped(data_type, readlen)
-                    target_buffer[mem_index] = re[buf_index]
-                else:
-                    f.seek(data_type.itemsize * readlen)
-            alen = f.get_raw_memmapped(util._head_type)
-            if alen != length:
-                raise IOError("Unexpected FORTRAN block length (tail) %d!=%d" % (alen, length))
+            h = {k: v for k, v in zip(genic_header['keys'], f.read_vector(genic_header['dtype'])[0])}
+            ii = 0
+            for _ in range(h['nz']):
```
Comment (apontzen) [inline pynbody/snapshot/grafic.py]: In its current form this breaks partial loading in Grafic (i.e. ability to load only part of the grid, which is provided by the LoadControl object)

Comment (cphyc) [reply pynbody/snapshot/grafic.py]: Actually the code as is seems to be working just fine, see the test grafic file.

Comment (apontzen) [reply pynbody/snapshot/grafic.py]: Unfortunately not - the test Grafic file is only 4^3=64, so the test as currently implemented always selects all particles!

If you update line 41 to `for test_len in [16, 32, 48]:`, for example, you see the problem

**Pull request merged**